modules/packages/Channel.chpl

```diff
@@ -0,0 +1,559 @@
+    */
+    pragma "no doc"
+    class Waiter {
+        type valueType;
+        var val : c_ptr(valueType);
+        var processPtr : c_ptr(single bool);
+        var isSelect : bool;
+        /* These variables are specific to the waiters for select statement*/
+        var isSelectDone : c_ptr(atomic int);
+        var selID : int = -1;
+
+        var prev : unmanaged Waiter(valueType)?;
+        var next : unmanaged Waiter(valueType)?;
+
+        proc init(ref value, ref process$ : single bool) {
+            valueType = value.type;
+            val = c_ptrTo(value);
+            processPtr = c_ptrTo(process$);
+            isSelect = false;
+        }
+
+        proc init(ref value, ref process$ : single bool, ref selDone : atomic int, caseID : int) {
+            valueType = value.eltType;
+            val = value;
+            processPtr = c_ptrTo(process$);
+            isSelect = true;
+            isSelectDone = c_ptrTo(selDone);
+            selID = caseID;
+        }
+
+        proc suspend() : bool {
+            return processPtr.deref().readFF();
+        }
+
+        proc release(status : bool) {
+            processPtr.deref().writeEF(status);
+        }
+    }
+
+    /*
+    Implementation of doubly-ended queue to keep track of suspended receiving
+    and sending tasks.
+    */
+    pragma "no doc"
+    class WaiterQue {
+        type eltType;
+        var front : unmanaged Waiter(eltType)?;
+        var back : unmanaged Waiter(eltType)?;
+
+        /* Push the `waiter` into the queue */
+        proc enque(waiter : unmanaged Waiter(eltType)) {
+            if front == nil {
+                front = waiter;
+                back = waiter;
+            }
+            else {
+                back!.next = waiter;
+                waiter.prev = back;
+                back = waiter;
+            }
+        }
+
+        proc isEmpty() : bool{
+            return (front == nil);
+        }
+
+        /* Pop the first value from the queue */
+        proc deque() : unmanaged Waiter(eltType) {
+            var waiter : unmanaged Waiter(eltType)?;
+            if front == nil {
+                // Error
+                writeln("Error");
+            }
+            else if front == back {
+                waiter = front;
+                front = nil;
+                back = nil;
+            }
+            else {
+                waiter = front;
+                front = front!.next;
+                front!.prev = nil;
+                waiter!.next = nil;
+            }
+            return waiter!;
+        }
+
+        /* Remove the specified entry from the queue */
+        proc deque(waiter : unmanaged Waiter(eltType)) {
+            if waiter == front {
+                deque();
+            }
+            else if waiter == back {
+                back = back!.prev;
+                back!.next = nil;
+            }
+            else if waiter.prev != nil && waiter.next != nil {
```

Comment:
This function is also used specifically by select statements. It may happen that the waiter is already dequed from the queue. In that case we don't have to do anything.

Reply:
Can you add a comment explaining that?

I would just add an

```chapel
else {
  // This function is also used specifically by select statements.
  // It may happen that the waiter is already dequed from the queue.
  // In that case we don't have to do anything.
}
```